File: server/room.c
/**
 * room.c - 房间管理模块实现
 */

#include "room.h"
#include "player.h"
#include "map.h"
#include "network.h"
#include "wal.h"
#include "recovery.h"
#include "snapshot.h"
#include "log.h"
#include "../common/protocol.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/* 全局状态 */
static Room *rooms[MAX_ROOMS];
static int next_room_id = 1;
static pthread_mutex_t rooms_mutex = PTHREAD_MUTEX_INITIALIZER;

/* 状态名称 */
static const char *status_names[] = {
    "WAITING",
    "STARTING",
    "GAMING",
    "ENDED"
};

int room_init(void) {
    pthread_mutex_lock(&rooms_mutex);
    
    for (int i = 0; i < MAX_ROOMS; i++) {
        rooms[i] = NULL;
    }
    
    /* 从 WAL 文件恢复最大房间 ID，避免 ID 冲突 */
    int max_id = recovery_get_max_room_id();
    next_room_id = max_id + 1;
    
    pthread_mutex_unlock(&rooms_mutex);
    
    log_info("Room manager initialized, next_room_id=%d", next_room_id);
    return 0;
}

void room_cleanup(void) {
    pthread_mutex_lock(&rooms_mutex);
    
    /* 第一步：通知所有游戏线程停止 */
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] != NULL) {
            pthread_mutex_lock(&rooms[i]->lock);
            rooms[i]->game_thread_running = 0;
            rooms[i]->status = ROOM_ENDED;
            pthread_mutex_unlock(&rooms[i]->lock);
        }
    }
    
    pthread_mutex_unlock(&rooms_mutex);
    
    /* 第二步：等待游戏线程退出（最多等待 2 秒） */
    usleep(200000);  /* 等待 200ms 让线程有机会退出 */
    
    pthread_mutex_lock(&rooms_mutex);
    
    /* 第三步：释放资源 */
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] != NULL) {
            pthread_mutex_destroy(&rooms[i]->lock);
            free(rooms[i]);
            rooms[i] = NULL;
        }
    }
    
    pthread_mutex_unlock(&rooms_mutex);
    
    log_info("Room manager cleaned up");
}

Room *room_create(const char *name, int max_players) {
    if (name == NULL || strlen(name) == 0) {
        log_error("room_create: name is NULL or empty");
        return NULL;
    }
    
    /* 验证人数范围 */
    if (max_players < MIN_ROOM_PLAYERS || max_players > MAX_ROOM_PLAYERS) {
        max_players = MAX_ROOM_PLAYERS;
    }
    
    pthread_mutex_lock(&rooms_mutex);
    
    /* 查找空槽位 */
    int slot = -1;
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] == NULL) {
            slot = i;
            break;
        }
    }
    
    if (slot < 0) {
        log_error("No available room slots");
        pthread_mutex_unlock(&rooms_mutex);
        return NULL;
    }
    
    /* 分配房间 */
    Room *room = calloc(1, sizeof(Room));
    if (room == NULL) {
        log_error("Failed to allocate room");
        pthread_mutex_unlock(&rooms_mutex);
        return NULL;
    }
    
    /* 初始化 */
    room->id = next_room_id++;
    strncpy(room->name, name, MAX_ROOM_NAME - 1);
    room->name[MAX_ROOM_NAME - 1] = '\0';
    room->max_players = max_players;
    room->player_count = 0;
    room->status = ROOM_WAITING;
    
    /* 初始化玩家列表 */
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        room->player_ids[i] = -1;
    }
    
    /* 地图数据 */
    room->item_count = 0;
    room->poison_radius = map_get_initial_poison_radius();
    
    /* 时间 */
    room->game_start_time = 0;
    room->last_item_spawn = 0;
    room->last_poison_shrink = 0;
    room->countdown_start = 0;
    
    /* 线程 */
    room->game_thread_running = 0;
    
    /* WAL */
    room->wal = NULL;
    
    /* 恢复相关 */
    room->is_recovery_room = 0;
    room->expected_players = 0;
    room->recovery_start_time = 0;
    room->original_room_id = -1;
    
    /* 快照相关 */
    room->last_snapshot_time = 0;
    
    /* 初始化锁 */
    pthread_mutex_init(&room->lock, NULL);
    
    rooms[slot] = room;
    
    pthread_mutex_unlock(&rooms_mutex);
    
    log_info("Room created: id=%d, name=%s, max=%d", room->id, name, max_players);
    return room;
}

void room_destroy(Room *room) {
    if (room == NULL) {
        return;
    }
    
    pthread_mutex_lock(&rooms_mutex);
    
    /* 从数组中移除 */
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] == room) {
            rooms[i] = NULL;
            break;
        }
    }
    
    int id = room->id;
    
    /* 关闭并删除 WAL 和快照（所有玩家退出，不需要恢复） */
    if (room->wal != NULL) {
        /* 恢复房间使用原始房间 ID 的 WAL 文件 */
        int wal_room_id = (room->original_room_id >= 0) ? room->original_room_id : id;
        wal_close_for_room(room->wal);
        room->wal = NULL;
        wal_delete_for_room(wal_room_id);
        snapshot_delete(wal_room_id);
        log_info("WAL and snapshot deleted for room %d (room destroyed)", wal_room_id);
    }
    
    pthread_mutex_destroy(&room->lock);
    free(room);
    
    pthread_mutex_unlock(&rooms_mutex);
    
    log_info("Room destroyed: id=%d", id);
}

Room *room_find_by_id(int id) {
    pthread_mutex_lock(&rooms_mutex);
    
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] != NULL && rooms[i]->id == id) {
            pthread_mutex_unlock(&rooms_mutex);
            return rooms[i];
        }
    }
    
    pthread_mutex_unlock(&rooms_mutex);
    return NULL;
}

int room_add_player(Room *room, Player *player) {
    if (room == NULL || player == NULL) {
        return -1;
    }
    
    pthread_mutex_lock(&room->lock);
    
    /* 检查房间状态 */
    if (room->status == ROOM_GAMING) {
        pthread_mutex_unlock(&room->lock);
        return -2;  /* 游戏进行中 */
    }
    
    /* 检查人数 */
    if (room->player_count >= room->max_players) {
        pthread_mutex_unlock(&room->lock);
        return -1;  /* 房间已满 */
    }
    
    /* 查找空位 */
    int slot = -1;
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] < 0) {
            slot = i;
            break;
        }
    }
    
    if (slot < 0) {
        pthread_mutex_unlock(&room->lock);
        return -1;
    }
    
    /* 添加玩家 */
    room->player_ids[slot] = player->id;
    room->player_count++;
    
    /* 更新玩家状态 */
    pthread_mutex_lock(&player->lock);
    player->room_id = room->id;
    player->status = PLAYER_IN_ROOM;
    pthread_mutex_unlock(&player->lock);
    
    pthread_mutex_unlock(&room->lock);
    
    log_info("Player %d joined room %d", player->id, room->id);
    
    /* 广播玩家加入 */
    char msg[MAX_MSG_LEN];
    protocol_build_player_join(msg, sizeof(msg), player->id, player->username);
    room_broadcast(room, msg);
    
    return 0;
}

int room_remove_player(Room *room, Player *player) {
    if (room == NULL || player == NULL) {
        return -1;
    }
    
    pthread_mutex_lock(&room->lock);
    
    /* 查找并移除玩家 */
    int found = 0;
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] == player->id) {
            room->player_ids[i] = -1;
            room->player_count--;
            found = 1;
            break;
        }
    }
    
    if (!found) {
        pthread_mutex_unlock(&room->lock);
        return -1;
    }
    
    /* 更新玩家状态 */
    pthread_mutex_lock(&player->lock);
    player->room_id = -1;
    player->status = PLAYER_LOBBY;
    pthread_mutex_unlock(&player->lock);
    
    int room_id = room->id;
    int player_id = player->id;
    int remaining = room->player_count;
    
    log_info("Player %d left room %d", player_id, room_id);
    
    /* 先释放锁，再广播和销毁 */
    pthread_mutex_unlock(&room->lock);
    
    /* 广播玩家离开（不持有 room->lock） */
    if (remaining > 0) {
        char msg[MAX_MSG_LEN];
        protocol_build_player_leave(msg, sizeof(msg), player_id);
        room_broadcast(room, msg);
    }
    
    /* 如果房间空了，销毁房间 */
    if (remaining == 0) {
        room_destroy(room);
    }
    
    return 0;
}

int room_start_game(Room *room) {
    if (room == NULL) {
        return -1;
    }
    
    int player_ids[MAX_ROOM_PLAYERS];
    int id_count = 0;
    
    pthread_mutex_lock(&room->lock);
    
    if (room->status != ROOM_WAITING && room->status != ROOM_STARTING) {
        pthread_mutex_unlock(&room->lock);
        return -1;
    }
    
    if (room->player_count < MIN_ROOM_PLAYERS) {
        pthread_mutex_unlock(&room->lock);
        return -1;
    }
    
    /* 生成地图 */
    map_generate(room->map);
    
    /* 初始化毒圈 */
    room->poison_radius = map_get_initial_poison_radius();
    
    /* 初始化道具 - 在 $ 刷新点生成初始道具 */
    room->item_count = 0;
    for (int y = 0; y < MAP_HEIGHT && room->item_count < MAX_MAP_ITEMS; y++) {
        for (int x = 0; x < MAP_WIDTH && room->item_count < MAX_MAP_ITEMS; x++) {
            if (room->map[y][x] == '$') {
                /* 随机生成道具类型 */
                MapItemType type = (MapItemType)(rand() % 3 + 1);
                room->items[room->item_count].x = x;
                room->items[room->item_count].y = y;
                room->items[room->item_count].type = type;
                room->items[room->item_count].active = 1;
                room->item_count++;
            }
        }
    }
    
    /* 设置时间 */
    room->game_start_time = player_get_time_ms();
    room->last_item_spawn = room->game_start_time;
    room->last_poison_shrink = room->game_start_time;
    
    /* 收集玩家 ID */
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] >= 0) {
            player_ids[id_count++] = room->player_ids[i];
        }
    }
    
    room->status = ROOM_GAMING;
    
    pthread_mutex_unlock(&room->lock);
    
    /* 初始化玩家位置和状态（不持有 room->lock） */
    for (int i = 0; i < id_count; i++) {
        Player *p = player_find_by_id(player_ids[i]);
        if (p != NULL) {
            player_reset_game_state(p);
            
            /* 随机位置 */
            int x, y;
            pthread_mutex_lock(&room->lock);
            map_random_position(room->map, &x, &y);
            pthread_mutex_unlock(&room->lock);
            
            pthread_mutex_lock(&p->lock);
            p->x = x;
            p->y = y;
            p->status = PLAYER_GAMING;
            pthread_mutex_unlock(&p->lock);
        }
    }
    
    log_info("Game started in room %d", room->id);
    
    /* 写入 WAL - 游戏开始 */
    if (room->wal != NULL) {
        char wal_data[512];
        snprintf(wal_data, sizeof(wal_data), "room_name=%s,max_players=%d",
                 room->name, room->max_players);
        wal_write(room->wal, WAL_GAME_START, wal_data);
        
        /* 写入每个玩家的初始状态 */
        for (int i = 0; i < id_count; i++) {
            Player *p = player_find_by_id(player_ids[i]);
            if (p != NULL) {
                pthread_mutex_lock(&p->lock);
                snprintf(wal_data, sizeof(wal_data),
                         "pid=%d,username=%s,x=%d,y=%d,hp=%d,max_hp=%d,atk=%d,def=%d,shield=%d,inv=%d,%d,%d,%d,%d",
                         p->id, p->username, p->x, p->y, p->hp, p->max_hp, p->atk, p->def,
                         p->has_shield,
                         (int)p->inventory[0].type, (int)p->inventory[1].type,
                         (int)p->inventory[2].type, (int)p->inventory[3].type,
                         (int)p->inventory[4].type);
                pthread_mutex_unlock(&p->lock);
                wal_write(room->wal, WAL_PLAYER_JOIN, wal_data);
            }
        }
        
        /* 写入初始道具状态 */
        pthread_mutex_lock(&room->lock);
        for (int i = 0; i < room->item_count; i++) {
            if (room->items[i].active) {
                snprintf(wal_data, sizeof(wal_data), "type=%d,x=%d,y=%d",
                         room->items[i].type, room->items[i].x, room->items[i].y);
                wal_write(room->wal, WAL_ITEM_SPAWN, wal_data);
            }
        }
        pthread_mutex_unlock(&room->lock);
        
        wal_sync(room->wal);
    }
    
    /* 广播游戏开始 */
    char msg[MAX_MSG_LEN];
    protocol_build_game_start(msg, sizeof(msg));
    room_broadcast(room, msg);
    
    /* 地图模板在客户端，不需要发送 */
    
    return 0;
}

void room_end_game(Room *room, int winner_id) {
    if (room == NULL) {
        return;
    }
    
    int player_ids[MAX_ROOM_PLAYERS];
    int count = 0;
    
    pthread_mutex_lock(&room->lock);
    room->status = ROOM_ENDED;
    room->game_thread_running = 0;
    
    /* 收集玩家 ID */
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] >= 0) {
            player_ids[count++] = room->player_ids[i];
        }
    }
    pthread_mutex_unlock(&room->lock);
    
    log_info("Game ended in room %d, winner=%d", room->id, winner_id);
    
    /* 写入 WAL - 游戏结束 */
    pthread_mutex_lock(&room->lock);
    if (room->wal != NULL) {
        char wal_data[64];
        snprintf(wal_data, sizeof(wal_data), "winner=%d", winner_id);
        wal_write(room->wal, WAL_GAME_END, wal_data);
        wal_sync(room->wal);
    }
    pthread_mutex_unlock(&room->lock);
    
    /* 广播游戏结束 */
    char msg[MAX_MSG_LEN];
    protocol_build_game_end(msg, sizeof(msg), winner_id, "");
    room_broadcast(room, msg);
    
    /* 重置玩家状态（不持有 room->lock） */
    for (int i = 0; i < count; i++) {
        Player *p = player_find_by_id(player_ids[i]);
        if (p != NULL) {
            pthread_mutex_lock(&p->lock);
            p->status = PLAYER_IN_ROOM;
            /* 重置游戏相关状态 */
            p->hp = p->max_hp;
            p->x = 0;
            p->y = 0;
            p->has_shield = 0;
            p->atk = p->base_atk;
            p->atk_buff_expire = 0;
            p->atk_buff_warned = 0;
            p->inventory_count = 0;
            pthread_mutex_unlock(&p->lock);
        }
    }
    
    /* 重置房间状态 */
    pthread_mutex_lock(&room->lock);
    room->status = ROOM_WAITING;
    
    /* 清空地图 - 下次开始时会重新生成 */
    memset(room->map, 0, sizeof(room->map));
    
    /* 清空道具 */
    room->item_count = 0;
    for (int i = 0; i < MAX_MAP_ITEMS; i++) {
        room->items[i].active = 0;
    }
    
    /* 重置毒圈 */
    room->poison_radius = map_get_initial_poison_radius();
    
    /* 重置时间 */
    room->game_start_time = 0;
    room->last_item_spawn = 0;
    room->last_poison_shrink = 0;
    
    /* 关闭 WAL 并删除文件（游戏正常结束，不需要恢复） */
    if (room->wal != NULL) {
        /* 恢复房间使用原始房间 ID 的 WAL 文件 */
        int wal_room_id = (room->original_room_id >= 0) ? room->original_room_id : room->id;
        wal_close_for_room(room->wal);
        room->wal = NULL;
        wal_delete_for_room(wal_room_id);
        snapshot_delete(wal_room_id);  /* 同时删除快照 */
        
        /* 删除 .recovery 文件 */
        char recovery_path[512];
        snprintf(recovery_path, sizeof(recovery_path), "%s/room_%d.recovery", WAL_DIR, wal_room_id);
        unlink(recovery_path);
        
        log_info("WAL, snapshot and recovery files deleted for room %d (game ended normally)", wal_room_id);
    }
    
    /* 重置恢复相关字段 */
    room->is_recovery_room = 0;
    room->expected_players = 0;
    room->recovery_start_time = 0;
    room->original_room_id = -1;
    room->last_snapshot_time = 0;
    
    pthread_mutex_unlock(&room->lock);
}

int room_broadcast(Room *room, const char *msg) {
    if (room == NULL || msg == NULL) {
        return 0;
    }
    
    int sent = 0;
    int player_ids[MAX_ROOM_PLAYERS];
    int count = 0;
    
    /* 先收集玩家 ID */
    pthread_mutex_lock(&room->lock);
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] >= 0) {
            player_ids[count++] = room->player_ids[i];
        }
    }
    pthread_mutex_unlock(&room->lock);
    
    /* 然后发送消息（不持有锁） */
    for (int i = 0; i < count; i++) {
        Player *p = player_find_by_id(player_ids[i]);
        if (p != NULL && player_send(p, msg) == 0) {
            sent++;
        }
    }
    
    return sent;
}

int room_get_list(char *buf, int max_len) {
    if (buf == NULL || max_len <= 0) {
        return 0;
    }
    
    /* 先收集房间信息 */
    struct {
        int id;
        char name[MAX_ROOM_NAME];
        int player_count;
        int max_players;
        int status;
    } room_info[MAX_ROOMS];
    int room_count = 0;
    
    pthread_mutex_lock(&rooms_mutex);
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] != NULL) {
            Room *r = rooms[i];
            pthread_mutex_lock(&r->lock);
            room_info[room_count].id = r->id;
            strncpy(room_info[room_count].name, r->name, MAX_ROOM_NAME - 1);
            room_info[room_count].name[MAX_ROOM_NAME - 1] = '\0';
            room_info[room_count].player_count = r->player_count;
            room_info[room_count].max_players = r->max_players;
            room_info[room_count].status = r->status;
            pthread_mutex_unlock(&r->lock);
            room_count++;
        }
    }
    pthread_mutex_unlock(&rooms_mutex);
    
    /* 构建响应字符串 */
    char data[MAX_MSG_LEN] = "";
    int pos = 0;
    int first = 1;
    
    for (int i = 0; i < room_count; i++) {
        int written = snprintf(data + pos, sizeof(data) - (size_t)pos,
                               "%s%d,%s,%d,%d,%d",
                               first ? "" : ";",
                               room_info[i].id, room_info[i].name,
                               room_info[i].player_count,
                               room_info[i].max_players, room_info[i].status);
        
        if (written > 0 && pos + written < (int)sizeof(data)) {
            pos += written;
            first = 0;
        }
    }
    
    return protocol_build_room_list(buf, max_len, data);
}

int room_all_ready(Room *room) {
    if (room == NULL) {
        return 0;
    }
    
    int player_ids[MAX_ROOM_PLAYERS];
    int count = 0;
    int player_count = 0;
    
    pthread_mutex_lock(&room->lock);
    player_count = room->player_count;
    if (player_count < MIN_ROOM_PLAYERS) {
        pthread_mutex_unlock(&room->lock);
        return 0;
    }
    
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] >= 0) {
            player_ids[count++] = room->player_ids[i];
        }
    }
    pthread_mutex_unlock(&room->lock);
    
    int all_ready = 1;
    for (int i = 0; i < count; i++) {
        Player *p = player_find_by_id(player_ids[i]);
        if (p != NULL) {
            pthread_mutex_lock(&p->lock);
            if (p->status != PLAYER_READY) {
                all_ready = 0;
            }
            pthread_mutex_unlock(&p->lock);
            
            if (!all_ready) break;
        }
    }
    
    return all_ready;
}

int room_get_alive_count(Room *room) {
    if (room == NULL) {
        return 0;
    }
    
    int player_ids[MAX_ROOM_PLAYERS];
    int id_count = 0;
    
    pthread_mutex_lock(&room->lock);
    for (int i = 0; i < MAX_ROOM_PLAYERS; i++) {
        if (room->player_ids[i] >= 0) {
            player_ids[id_count++] = room->player_ids[i];
        }
    }
    pthread_mutex_unlock(&room->lock);
    
    int count = 0;
    for (int i = 0; i < id_count; i++) {
        Player *p = player_find_by_id(player_ids[i]);
        if (p != NULL) {
            pthread_mutex_lock(&p->lock);
            if (p->status == PLAYER_GAMING && p->hp > 0) {
                count++;
            }
            pthread_mutex_unlock(&p->lock);
        }
    }
    
    return count;
}

int room_get_count(void) {
    int count = 0;
    
    pthread_mutex_lock(&rooms_mutex);
    
    for (int i = 0; i < MAX_ROOMS; i++) {
        if (rooms[i] != NULL) {
            count++;
        }
    }
    
    pthread_mutex_unlock(&rooms_mutex);
    return count;
}

const char *room_status_name(RoomStatus status) {
    if (status >= 0 && status <= ROOM_ENDED) {
        return status_names[status];
    }
    return "UNKNOWN";
}

File: asciigame-go/internal/server/room.go
package server

import (
	"fmt"
	"log"
	"math/rand/v2"
	"sync"

	"github.com/heartlazyli/asciigame/internal/config"
	"github.com/heartlazyli/asciigame/internal/protocol"
)

// RoomStatus mirrors the C RoomStatus enum (room.h:19-24).
type RoomStatus int

const (
	RoomWaiting RoomStatus = iota
	RoomStarting
	RoomGaming
	RoomEnded
)

// mapItem mirrors the C MapItem (map.h:34-38), extended with spawn time for
// expiration.
type mapItem struct {
	x, y      int
	typ       ItemType
	active    bool
	spawnTime int64 // millisecond timestamp when this item appeared
}

// Room mirrors the C Room (room.h:27-63). Fields are guarded by mu. The
// per-room game loop goroutine and connection goroutines both touch these
// fields, exactly as the C game_thread and handler threads did.
type Room struct {
	srv        *Server
	id         int
	name       string
	maxPlayers int

	mu          sync.Mutex
	members     [config.MaxRoomPlayers]*Player // nil == empty slot
	playerCount int
	status      RoomStatus
	m           gameMap
	items       [config.MaxMapItems]mapItem
	itemCount   int
	poisonRadius int

	gameStartTime    int64
	lastItemSpawn    int64
	lastPoisonShrink int64
	lastSnapshotTime int64
	running          bool

	// lastStateSig is the marshaled GameState signature (timestamp excluded)
	// from the last broadcastState; used to skip identical broadcasts. Reset on
	// game start/end so the first state of each game always sends.
	lastStateSig []byte

	// Recovery bookkeeping (recovery.c). When a room is reconstructed after a
	// crash it waits for the expected players to reconnect before judging the
	// winner.
	isRecovery      bool
	expectedPlayers int
	recoveryStart   int64
	originalRoomID  int

	wal *wal
}

// createRoom mirrors room_create (room.c:83-166); max_players out of range is
// clamped to MaxRoomPlayers.
func (s *Server) createRoom(name string, maxPlayers int) *Room {
	if name == "" {
		return nil
	}
	if maxPlayers < config.MinRoomPlayers || maxPlayers > config.MaxRoomPlayers {
		maxPlayers = config.MaxRoomPlayers
	}
	s.rmu.Lock()
	defer s.rmu.Unlock()
	if len(s.rooms) >= config.MaxRooms {
		return nil
	}
	r := &Room{
		srv:            s,
		id:             s.nextRoomID,
		name:           truncate(name, config.MaxRoomName-1),
		maxPlayers:     maxPlayers,
		status:         RoomWaiting,
		poisonRadius:   mapInitialPoisonRadius(),
		originalRoomID: -1,
	}
	s.nextRoomID++
	s.rooms[r.id] = r
	log.Printf("room created: id=%d name=%s max=%d", r.id, name, maxPlayers)
	return r
}

// snapshotMembers returns the room's live players under the room lock, so
// callers can iterate without holding the lock or re-looking-up by id.
func (r *Room) snapshotMembers() []*Player {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.membersLocked()
}

// membersLocked returns the live players; the caller must hold r.mu.
func (r *Room) membersLocked() []*Player {
	out := make([]*Player, 0, r.playerCount)
	for _, p := range r.members {
		if p != nil {
			out = append(out, p)
		}
	}
	return out
}

// findRoomByID mirrors room_find_by_id (leaf lock).
func (s *Server) findRoomByID(id int) *Room {
	s.rmu.RLock()
	defer s.rmu.RUnlock()
	return s.rooms[id]
}

// destroyRoom mirrors room_destroy (room.c:168-202): stop the game loop, drop
// from the registry, close the WAL.
func (s *Server) destroyRoom(r *Room) {
	s.rmu.Lock()
	delete(s.rooms, r.id)
	s.rmu.Unlock()

	r.mu.Lock()
	r.running = false
	r.status = RoomEnded
	r.wal.close()
	r.wal = nil
	r.mu.Unlock()
	log.Printf("room destroyed: id=%d", r.id)
}

// addPlayer mirrors room_add_player (room.c:218-271):
//
//	0 ok, -1 full/no slot, -2 game in progress.
func (r *Room) addPlayer(p *Player) int {
	r.mu.Lock()
	if r.status == RoomGaming {
		r.mu.Unlock()
		return -2
	}
	if r.playerCount >= r.maxPlayers {
		r.mu.Unlock()
		return -1
	}
	slot := -1
	for i := 0; i < config.MaxRoomPlayers; i++ {
		if r.members[i] == nil {
			slot = i
			break
		}
	}
	if slot < 0 {
		r.mu.Unlock()
		return -1
	}
	r.members[slot] = p
	r.playerCount++
	r.mu.Unlock()

	p.mu.Lock()
	p.roomID = r.id
	p.status = StatusInRoom
	p.mu.Unlock()

	log.Printf("player %s joined room %d", p.label(), r.id)
	r.broadcast(protocol.NewPlayerJoin(int32(p.id), p.username))
	return 0
}

// removePlayer mirrors room_remove_player (room.c:273-324): remove, set player
// to lobby, broadcast leave, and destroy the room if it becomes empty.
func (r *Room) removePlayer(p *Player) {
	r.mu.Lock()
	found := false
	for i := 0; i < config.MaxRoomPlayers; i++ {
		if r.members[i] != nil && r.members[i].id == p.id {
			r.members[i] = nil
			r.playerCount--
			found = true
			break
		}
	}
	if !found {
		r.mu.Unlock()
		return
	}
	remaining := r.playerCount
	r.mu.Unlock()

	p.mu.Lock()
	p.roomID = -1
	p.status = StatusLobby
	p.mu.Unlock()

	if remaining > 0 {
		r.broadcast(protocol.NewPlayerLeave(int32(p.id)))
	} else {
		r.srv.destroyRoom(r)
	}
}

// broadcast mirrors room_broadcast (room.c:558-585): snapshot members under the
// room lock, then send off-lock so no socket write happens under a mutex.
func (r *Room) broadcast(f *protocol.Frame) {
	for _, p := range r.snapshotMembers() {
		p.Send(f)
	}
}

// allReady mirrors room_all_ready (room.c:641-679): at least MinRoomPlayers and
// every member in StatusReady.
func (r *Room) allReady() bool {
	r.mu.Lock()
	if r.playerCount < config.MinRoomPlayers {
		r.mu.Unlock()
		return false
	}
	members := r.membersLocked()
	r.mu.Unlock()

	for _, p := range members {
		p.mu.Lock()
		ready := p.status == StatusReady
		p.mu.Unlock()
		if !ready {
			return false
		}
	}
	return true
}

// startGame mirrors room_start_game (room.c:326-452): generate the map, seed
// items at spawn points, reset and place players, then broadcast GAME_START.
func (r *Room) startGame() int {
	r.mu.Lock()
	if r.status != RoomWaiting && r.status != RoomStarting {
		r.mu.Unlock()
		return -1
	}
	if r.playerCount < config.MinRoomPlayers {
		r.mu.Unlock()
		return -1
	}

	mapGenerate(&r.m)
	r.poisonRadius = mapInitialPoisonRadius()

	now := nowMS()
	r.itemCount = 0
	for y := 0; y < config.MapHeight && r.itemCount < config.MaxMapItems; y++ {
		for x := 0; x < config.MapWidth && r.itemCount < config.MaxMapItems; x++ {
			if r.m[y][x] == cellSpawn {
				r.items[r.itemCount] = mapItem{x: x, y: y, typ: ItemType(rand.IntN(3) + 1), active: true, spawnTime: now}
				r.itemCount++
			}
		}
	}

	r.gameStartTime = now
	r.lastItemSpawn = now
	r.lastPoisonShrink = now

	ids := r.membersLocked()
	r.status = RoomGaming
	r.lastStateSig = nil // first broadcastState of the game always sends
	r.mu.Unlock()

	// Reset and place players (not holding room lock while touching players,
	// except the brief map read under room lock, matching C).
	for _, p := range ids {
		p.resetGameState()
		r.mu.Lock()
		x, y := mapRandomPosition(&r.m)
		r.mu.Unlock()
		p.mu.Lock()
		p.x = x
		p.y = y
		p.status = StatusGaming
		p.mu.Unlock()
	}

	// Persist initial game state to the WAL (room_start_game, room.c:406-442).
	r.wal.write(walGameStart, fmt.Sprintf("room_name=%s,max_players=%d", r.name, r.maxPlayers))
	for _, p := range ids {
		p.mu.Lock()
		rec := fmt.Sprintf("pid=%d,username=%s,x=%d,y=%d,hp=%d,max_hp=%d,atk=%d,def=%d,shield=%d,inv=%d,%d,%d,%d,%d",
			p.id, p.username, p.x, p.y, p.hp, p.maxHP, p.atk, p.def, boolToInt(p.hasShield),
			int(p.inventory[0]), int(p.inventory[1]), int(p.inventory[2]), int(p.inventory[3]), int(p.inventory[4]))
		p.mu.Unlock()
		r.wal.write(walPlayerJoin, rec)
	}
	r.mu.Lock()
	for i := 0; i < r.itemCount; i++ {
		if r.items[i].active {
			r.wal.write(walItemSpawn, fmt.Sprintf("type=%d,x=%d,y=%d", int(r.items[i].typ), r.items[i].x, r.items[i].y))
		}
	}
	r.mu.Unlock()
	r.wal.sync()

	log.Printf("game started in room %d", r.id)
	r.broadcast(protocol.NewGameStart())
	return 0
}

// getRoomList mirrors room_get_list (room.c:587-639), returning a ROOM_LIST
// frame carrying one RoomInfo per live room.
func (s *Server) getRoomList() *protocol.Frame {
	s.rmu.RLock()
	rooms := make([]*Room, 0, len(s.rooms))
	for _, r := range s.rooms {
		rooms = append(rooms, r)
	}
	s.rmu.RUnlock()

	infos := make([]*protocol.RoomInfo, 0, len(rooms))
	for _, r := range rooms {
		r.mu.Lock()
		infos = append(infos, &protocol.RoomInfo{
			RoomId: int32(r.id), Name: r.name,
			PlayerCount: int32(r.playerCount), MaxPlayers: int32(r.maxPlayers),
			Status: int32(r.status),
		})
		r.mu.Unlock()
	}
	return protocol.NewRoomList(infos)
}

func truncate(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}
